doc/users-guide/users-guide-ipsec.adoc

```diff
@@ -0,0 +1,437 @@
+wish to use time-based SA limits may do so on their own using the timing
+facilities that ODP provides. However, since especially with inline IPsec
+processing, the application may not have explicit knowledge of the traffic
+volumes associated with a given SA, support for usage-based limits is
+integrated into ODP IPsec support.
+
+At `odp_ipsec_sa_create()` time, one of the fields in the
+`odp_ipsec_sa_param_t` struct is the `odp_ipsec_lifetime_t` sub-structure.
+This struct allows hard and/or soft limits to be specified in terms of total
+bytes encrypted/decrypted, total packet count, or both. A limit specification
+of 0 indicates no limit for that metric. If either is specified, the limit
+is triggered on whichever occurs first. Given the defined behavior of hard vs.
+soft limits, the soft limits, if used, should always be specified as lower
+than the hard limits. These should be sufficiently lower to enable adequate
+time to switch over to a replacement SA before the hard limit is reached.
+
+As noted, when an SA hard limit is reached the SA immediately enters the
+expired state and attempts to use it further are failed with an
+`odp_ipsec_result_t` that indicates a hard expiration limit. When a soft
+limit is reached for packets sent via `odp_ipsec_out_inline()`, this results
+in an `ODP_EVENT_IPSEC_STATUS` event being sent to the application on the
+queue associated with the SA that has reached the soft limit. This status
+event has an `odp_ipsec_status_id_t` of `ODP_IPSEC_STATUS_WARN` with a
+`odp_ipsec_warn_t` bits set to indicate the type of soft expiration reached.
+Receipt of this event alerts the application that the SA is nearing the end of
+its useful life and that it should be replaced. It is the application's
+responsibility to heed this warning. Soft limit alerts are edge triggered and
+are only issued once rather than repeatedly as the SA runs beyond the soft
+limit.
+
+When operating in lookaside mode, expiration limits are carried as a warning
+in the `odp_op_status_t` section of the `odp_ipsec_result_t` struct. The same
+is true for inline inbound packets.
+
+==== SA Disablement and Destruction
+When it is time to retire an SA, the application does so by issuing a call
+to the `odp_ipsec_sa_destroy()` API. However, to better support graceful SA
+termination processing, ODP also provides the `odp_ipsec_sa_disable()` API.
+
+This call initiates termination processing for an SA by stopping use of the SA
+for new operations while still allowing those that are "in flight" to complete
+processing. All newly arriving packets following this call will carry a
+disabled warning in the `odp_op_status_t` section of the `odp_ipsec_result_t`
+struct. So after disabling the SA, the application can process pending packets
+normally until it sees packets marked with the disabled warning. At this point
+it knows that all pending packets that arrived before the disable have been
```

Comment:
Well, we do not have #197 merged, so it is too early to depend on that.
Also, frankly speaking, this `sa_disabled` warning inside `odp_ipsec_result_t` is a backup plan. It is expected that most of the implementations will report this as a proper `ODP_IPSEC_STATUS` event, carrying this warning bit inside.

Reply:
I'd like to see us all make it a priority to close on #197 next week. I'll post a doc update once we're all agreed on the precise semantics and operation flow. So I'm fine with holding off merging this PR until then.